### ISSUE TITLE
Merge duplicate function for persons

### ISSUE THREAD

**Issue by kwa20** [SORMAS-Foundation/SORMAS-Project#5576]
### Feature Description

Thus far, the option to merge duplicates is only available for cases and contacts. Often enough however, persons are created as duplicates due to limit person duplicate suggestions and there is no option to merely merge persons that exist as duplicates in the database. This is very relevant because users are predominantly working on a person basis and having duplicates in the system can make this more difficult.   

This issue is further part of a larger concept issue: #3560 

### Proposed Change

Introduce a merge duplicate option for persons. This merge duplicate function should have its separate view (as seen for cases and contacts) and include the following options:

Note: In order to merge persons, the persons table might need a deleted column in order to be able to discard duplicates.

- Add a button to navigate back to the person directory.
- Add a button to open a merge duplicates guide for persons. This guide should explain how the duplicate detection for persons works and how merging of duplicates can be done.
- Add filters:
    - creation date time period filters (date filters)
    - birthdate filters (drop-downs)
    - free text filter for names, id, phone number, email, etc. (as seen on person directory)
    - present condition (drop-down)
    - region, district, community - home address (drop-downs)
    - Filter to show persons with differing birthdates (checkbox)
    - Filter to show persons with different sexes (checkbox)

- Add a list of suggested duplicates:
[_This list would be handled differently to how cases and contacts are handled due to performance implications and other limitations. The idea is to not having to compare every person with all other persons in the system every time the view loads._]
    - Add a database table for person duplicate suggestions. In this table, store the suggested duplicates that come up during the creation of entities or imports. Saved should only be the suggestions that actually might have led to the creation of potential duplicates, i.e. selecting a suggested person during the creation of e.g. a case will not produce an entry; if a person was suggested but the user created a new person, an entry will be stored.
    - When person data relevant for duplicate detection is changed and saved, the saved person should get compared with all other persons in the system again and now check if there is a potential duplicate after these changes.
        - If potential duplicates were found, prompt the user with a pop-up and inform them that the person is now similar to an existing person in the system. Give them the option to either merge, pick (with one of the duplicates as the leading person) or proceed without any action. If the user proceeded without any action, the suggestion will be stored in the table.
    - On the merge view, display the list of suggested duplicates.
    - Creation date filters should be based on the duplicate suggestions.
    - Personal filters should take into consideration any person that is listed under the suggested duplicates.

- Add action buttons similar to other merge duplicate function to either
    - Merge (persons should be merged with the leading person adopting data from the one to be discarded)
    - Pick (persons should be picked and not adopt any additional data from the one to be discarded)
    - Discard (this would discard the person duplicate suggestions in the table and remove them completely from the list)

**Merging Process Additional Information**
- Adopt entities by linking them to the remaining person upon discarding the other one.
- Follow-up visits from discarded entities need to be adopted.
- personContactDetails: 
    - If the contact information is identical, do not add information from person to be discarded.
    - If person to be discarded has contact information that is not present in the remaining person, add to the list.
    - If contact information that will be adopted are tagged as primary in the person to be discarded, untag the primary label if primary information for the type already exists in the remaining person, otherwise leave the information tagged as primary.
- If the person is registered in an external symptom journal, cancel the external follow-up upon discarding the person.
- addresses:
    - If address information is identical, do not add information from person to be discarded.
    - If person to be discarded has address information that is not present in the remaining person, add to the list.

**Merging of associated entities**
When persons are being merged and entities adopted in the remaining person, there might be some duplicate cases or contacts (since event participants are person-based, this might not be a problem for them).
- Upon a person merge, compare the entities of the same type of this person and if duplicate entities might exist, suggest them in a pop-up window. 
- This pop-up window should have separate lists for cases and contacts with options to **merge**, **pick** and **undo** per suggested duplicate. These selections should only be pre-selections. Undo is necessary for not having to discard the whole action if merging or picking were mistakenly chosen.
- It should also have buttons to discard or confirm. Discarding will not merge or pick any suggested entities. Confirming will merge or pick the entities that were marked accordingly.

### Additional Information

(might diverge from the description above)
<img width="769" alt="personDuplicateMerge2" src="https://user-images.githubusercontent.com/70317594/121207249-6d94d300-c879-11eb-821c-32e5afb0a100.PNG">

**Comment by kwa20** [comment]
@Jan-Boehme @Candice-Louw @Marko-ilmkreis @Johannes-Schmidt-GA-RT Feel free to provide feedback in case I've missed anything or reference more contributors.

@Jan-Boehme If we want to rework the person duplicate detection logic, we should probably do that in a separate issue that goes in parallel to this one. Same goes for a potential force merge bulk action option. I'll also create a separate issue for configurations associated to duplicate parameters and link it to this one.

**Comment by bernardsilenou** [comment]
@kwa20  
I have comments, whats your thought on it pls:
- if a person is duplicate in the system, then there is a duplicate entity (case, contact, ep) for that person and each of these entities can be more than one. By merging the entities of a person, we indirectly merge persons ie if there are no duplicate entities, then we would not have duplicate persons.
- If we merge persons, then what should happen to their entities ? I suppose we have to merge them also or link them to the resulting person with validation by the user. This is the point that is not still clear on how we would do it.
- If we decide not to merge or reassign all the entities to the resulting person (since the other person would be deleted as a result of the person merge), then we would end up with non duplicate persons but still duplicate entities with some linked to deleted persons.
- So all boils down to duplicate entities detection, maybe we need to instead refine the duplicate entities detection which would indirectly resolve the duplicate person  issue.

**Comment by kwa20** [comment]
@bernardsilenou I get your point, however, there can be many ways in which we can have duplicates without having duplicate entities. For example:
- A person had many contacts to different source cases. All contacts describe a separate contact while all of them belong to the same person that has been entered multiple times.
- Same would be possible for event participants.
- If a person has an initial case and is/was reinfected, the person can be a duplicate while the cases are not.

I'll update the initial description but the main idea would be to link the entities of the person to be discarded to the remaining person. We could then add an entity duplicate suggestion upon the person merge that suggests associated entities duplicates which merely compared entities of the same type to the same person.

**Comment by Marko-ilmkreis** [comment]
@bernardsilenou I fully agree with @kwa20 as we have a lot, probably more then 500 persons as a duplicate. Each with an own contact, case or ep, which is not a duplicate. This results in then not very good duplicate detection in the past. Sometimes also misspelling the names (just swap a letter). If that is corrected in the future it also stays as a duplicate person in Sormas.

**Comment by bernardsilenou** [comment]
@kwa20 @Marko-ilmkreis  
I think my main point has not still been address. It is based on the statement below:
`Merge (persons should be merged with the leading person adopting data from the one to be discarded)`
Q1: Does this mean we would merge all entities or transfer to the new person automatically? 

`Pick (persons should be picked and not adopt any additional data from the one to be discarded)`

Q2. If we pick, then are we saying that the deleted person and all associated entities should be deleted also? If no, then system would reject this decision because we can not have in deleted entities pointing to deleted persons in the system. If no then we end up with the same duplicate persons as before.

`Hide (hiding duplicates should not only temporarily but rather store them long term under hidden duplicates. This is so that persons that are wrongfully suggested as duplicates don't reappear again every time the view is loaded again. If feasable, this should also be added to cases and contacts merge duplicate views)`
This sounds like the soft deletion of persons

**Comment by kwa20** [comment]
@bernardsilenou 
Q1: Yes, all associated entities would be transferred to the remaining person.
Q2: Yes, all associated entities will be deleted alongside the discarded person.
Q3: Hiding is only associated with the display of suggested duplicates and independent of any deletions or entities itself.

**Comment by bernardsilenou** [comment]
@kwa20 I get you now. 
It is a very rough way of merging entities by merging persons. 
- You remember that when merging cases, we suggest the case to keep based on the completeness. So when merging persons, then we would need some information/ logic to help the user decide which of the persons to keep. If user pick a person linked to an ep while the other person is linked to cases and contacts, then all entities of the second person are deleted for good. 
- Maybe we should keep merge option only
- We can discuss in details tomorrow

**Comment by bernardsilenou** [comment]
@kwa20 @Marko-ilmkreis 
Questions please:
The merge duplicates functions for cases and contacts were added because the duplicate feature may not work all the time during entities creation. This often happen with mobile app when user is offline.
 If users run merge duplicates for cases, contacts and ep (assume we have this also), would this not clear all duplicate persons in their person directory?
In addition, our ultimate goal should be to have non-duplicate entities, right? So if we go through the person or entities, at the end of the process we should end up having non duplicate entities.

**Comment by kwa20** [comment]
@bernardsilenou It wouldn't because of the same occurrences outlined in https://github.com/hzi-braunschweig/SORMAS-Project/issues/5576#issuecomment-850784649

**Comment by Candice-Louw** [comment]
Further input/comments/thumbs-up in support most welcome @Lucas-Walter-GA-Tuebingen @Tobiasschellhorn @TheGaHTK @heineltp @CriSi315 @OrgaLandkreisTuebingen, please (and referring someone else to comment, of course too).

If no response till end of the week (04.062021), we will assume that what has been defined here is complete and will sufficiently address your needs.

**Comment by Marko-ilmkreis** [comment]
> @kwa20 @Marko-ilmkreis 
> Questions please:
> The merge duplicates functions for cases and contacts were added because the duplicate feature may not work all the time during entities creation. This often happen with mobile app when user is offline.
>  If users run merge duplicates for cases, contacts and ep (assume we have this also), would this not clear all duplicate persons in their person directory?
> In addition, our ultimate goal should be to have non-duplicate entities, right? So if we go through the person or entities, at the end of the process we should end up having non duplicate entities.

The duplicate detection for cases makes only sense if you have a duplicate case. If you have two unique cases(e.g. One non-case and one confirmed case) , but linked to two person entities of the same real person there must be an option to identify the duplicate person and merge the person only.

**Comment by kwa20** [comment]
@ChristopherRiedel @Jan-Boehme I've updated the initial description according to the discussion last Friday. Please check the section underneath:
> Add a list of suggested duplicates:

One major downside I thought about again was that mainly listing the suggestions automatically disregards all potential duplicates that fell through the initial detection upon entity creation. We might want to take this into consideration for the final refinement.

**Comment by Marko-ilmkreis** [comment]
@kwa20 @ChristopherRiedel  @Jan-Boehme

Thanks for your great work. Unfortunately I was off last Friday and have another meeting outside the office this Friday. :-/

> * When person data relevant for duplicate detection is changed and saved, the saved person should get compared with all other persons in the system again and now check if there is a potential duplicate after these changes.

Specially this would be very helpful in my mind.

Whats with the current duplicates in the system. Will there be an initial run of the duplicate detection for persons to fill the table?

**Comment by heineltp** [comment]
Good morning,

i am out of the sormas work since easter and had no chance to follow github so i just want to give a very important information that i does not get lost.

One example:
After our initial import we had the problem that one person had two different cases.
So: Same person, two totaly different self standing cases e.g. march 2020 and august 2020

In the duplicitycheck at the import we said: create new case.
In the duplicitycheck after the import the system still thinks that these two cases are the same.

So in this feature please don't forget to give the user the chance to set an "ignore in the future" checkbox.



And then in addition:
Because of the sormas structure we had the problem that we had Person dublicates after the import.
So we had two person Identities each with an case. These cases are totaly selfstanding like the example above.
The user also needs the possibility to bring the person dublicate to one person together and KEEP BOTH cases.


Best wishes for your work.

Greetings
Philippe

**Comment by bernardsilenou** [comment]
@kwa20  This comment https://github.com/hzi-braunschweig/SORMAS-Project/issues/5576#issuecomment-857380631 is the same as whet we discussed yesterday.
- to reproduce this bug is to export the same case, delete uuid, edit report date and onset date and re-import the same data

**Comment by StefanKock** [comment]
@MateStrysewske will look into this.

**Comment by MateStrysewske** [comment]
@kwa20 @JaquM @bernardsilenou The currently proposed solution regarding the way duplicates are retrieved has some major flaws in my eyes. I understand that performance was the main reasoning behind it, but by doing it this way, we:

a) overlook potentially duplicate persons that have been created in the past (because we have no records for them)
b) overlook potentially duplicate persons that have been pushed into the system via ReST because there is no duplicate detection in place there right now

Both would significantly reduce the usefulness of this tool.

We already changed the merge duplicate cases view in a way that upon initially entering the view, no duplicates are shown and you instead are encouraged to specify relatively strict filters first before calculating them. If we would do the same for persons, maybe that would be enough?

**Comment by bernardsilenou** [comment]
@MateStrysewske 
- This is  one good way to go, not only for persons but we can also extend it to cases and other entities. To build a table of possible duplicates each night that meet the similarity threshold.
- Do we initially check for all person in the db and  subsequent update the entries based on  persons updated in the previous day only?

**Comment by MateStrysewske** [comment]
@bernardsilenou That would be how it's working, yes.

**Comment by kwa20** [comment]
@MateStrysewske @JaquM I'd suggest to turn this issue into an epic and review the content and split it into seperate issues. We can revise the initial description in regards to the duplicate detection method.
